rama-core/src/matcher/ext.rs

```diff
@@ -0,0 +1,78 @@
+use crate::{
+    matcher::{Extensions, Matcher},
+    Context,
+};
+use std::any::Any;
+
+pub struct ExtensionMatcher<T: Send + Sync> {
+    marker: T,
+}
+
+struct Const<T: Send + Sync>(T);
+struct FnBox(Box<dyn Fn(&dyn Any) -> bool + Send + Sync>);
+
+impl ExtensionMatcher<FnBox> {
+    pub fn with<T: Send + Sync + 'static, F: Fn(&T) -> bool + Send + Sync + 'static>(f: F) -> Self {
+        let wrapped_fn = move |v: &dyn Any| {
+            if let Some(concrete) = v.downcast_ref::<T>() {
+                f(concrete)
+            } else {
+                false
+            }
+        };
+        Self {
+            marker: FnBox(Box::new(wrapped_fn)),
+        }
+    }
+}
+
+impl<T: Send + Sync + Sync + std::cmp::PartialEq> ExtensionMatcher<Const<T>> {
+    pub fn constant(value: T) -> Self {
+        Self {
+            marker: Const(value),
+        }
+    }
+}
+
+impl<T: Send + Sync + std::cmp::PartialEq + 'static, State, Request> Matcher<State, Request>
+    for ExtensionMatcher<Const<T>>
+{
+    fn matches(&self, _ext: Option<&mut Extensions>, ctx: &Context<State>, _req: &Request) -> bool {
+        ctx.get::<T>()
+            .map(|v| v == &self.marker.0)
+            .unwrap_or_default()
+    }
+}
+
+impl<State, Request> Matcher<State, Request> for ExtensionMatcher<FnBox> {
+    fn matches(&self, _ext: Option<&mut Extensions>, ctx: &Context<State>, _req: &Request) -> bool {
+        ctx.get::<FnBox>()
+            .map(|v| (self.marker.0)(v))
+            .unwrap_or_default()
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
This is how I Would implement the tests, I left 2 todo's in there, as to allow you to add some test cases of your own:

```rust
#[cfg(test)]
mod test {
    use super::*;

    #[derive(Debug, Clone, PartialEq, Eq)]
    struct MyMarker(i32);

    #[test]
    fn test_extension_matcher() {
        let matcher = ExtensionMatcher::with_const(MyMarker(10));
        let mut ctx = Context::default();

        assert!(!matcher.matches(None, &ctx, &()));

        ctx.insert(MyMarker(10));
        assert!(matcher.matches(None, &ctx, &()));

        // TODO: test for 2 more cases or so :)
    }

    #[test]
    fn test_fn_extension_matcher() {
        let matcher = ExtensionMatcher::with_fn(|v: &MyMarker| v.0 % 2 == 0);
        let mut ctx = Context::default();

        assert!(!matcher.matches(None, &ctx, &()));

        ctx.insert(MyMarker(4));
        assert!(matcher.matches(None, &ctx, &()));

        // TODO: test for 2 more cases or so :)
    }
}
```